File: sdks/py/src/jeffs_brain_memory/memory/procedural.py
# SPDX-License-Identifier: Apache-2.0
"""Procedural memory detection from conversation transcripts."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from datetime import datetime, timezone

from ..llm.types import Role
from .types import Message


@dataclass(slots=True)
class ProceduralRecord:
    tier: str = ""
    name: str = ""
    task_context: str = ""
    outcome: str = ""
    observed_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    tool_calls: list[str] = field(default_factory=list)
    tags: list[str] = field(default_factory=list)


def detect_procedurals(messages: list[Message]) -> list[ProceduralRecord]:
    records: list[ProceduralRecord] = []
    now = datetime.now(timezone.utc)

    for s in _detect_skill_invocations(messages):
        records.append(
            ProceduralRecord(
                tier="skill",
                name=s["name"],
                task_context=s["context"],
                outcome=s["outcome"],
                observed_at=now,
                tool_calls=["skill"],
                tags=["procedural", "skill", s["name"]],
            )
        )
    for a in _detect_agent_invocations(messages):
        records.append(
            ProceduralRecord(
                tier="agent",
                name=a["name"],
                task_context=a["context"],
                outcome=a["outcome"],
                observed_at=now,
                tool_calls=["agent"],
                tags=["procedural", "agent", a["name"]],
            )
        )
    return records


def _detect_skill_invocations(messages: list[Message]) -> list[dict]:
    out: list[dict] = []
    for i, m in enumerate(messages):
        if m.role != Role.ASSISTANT:
            continue
        for tc in m.tool_calls:
            if tc.name != "skill":
                continue
            try:
                args = json.loads(tc.arguments)
            except json.JSONDecodeError:
                continue
            skill = args.get("skill", "") if isinstance(args, dict) else ""
            if not skill:
                continue
            outcome = infer_tool_call_outcome(messages, i, tc.id)
            ctx = infer_procedural_context(messages, i)
            out.append({"name": skill, "context": ctx, "outcome": outcome})
    return out


def _detect_agent_invocations(messages: list[Message]) -> list[dict]:
    out: list[dict] = []
    for i, m in enumerate(messages):
        if m.role != Role.ASSISTANT:
            continue
        for tc in m.tool_calls:
            if tc.name != "agent":
                continue
            try:
                args = json.loads(tc.arguments)
            except json.JSONDecodeError:
                continue
            agent_type = args.get("type", "") if isinstance(args, dict) else ""
            if not agent_type:
                continue
            outcome = infer_tool_call_outcome(messages, i, tc.id)
            prompt = args.get("prompt", "") if isinstance(args, dict) else ""
            ctx = prompt if prompt else ""
            if ctx and len(ctx) > 160:
                ctx = ctx[:160]
            if not ctx:
                ctx = infer_procedural_context(messages, i)
            out.append({"name": agent_type, "context": ctx, "outcome": outcome})
    return out


def infer_tool_call_outcome(
    messages: list[Message], after_index: int, tool_call_id: str
) -> str:
    for m in messages[after_index:]:
        if m.role != Role.TOOL:
            continue
        if m.tool_call_id == tool_call_id or (not tool_call_id and m.name == "skill"):
            content = m.content.lower()
            if "error" in content or "failed" in content:
                return "error"
            return "ok"
    return "partial"


def infer_procedural_context(messages: list[Message], before_index: int) -> str:
    for i in range(before_index - 1, -1, -1):
        if messages[i].role == Role.USER:
            ctx = messages[i].content
            if len(ctx) > 160:
                ctx = ctx[:160]
            return ctx
    return ""


def format_procedural_record(r: ProceduralRecord) -> str:
    parts: list[str] = ["---", f"name: {r.name}"]
    parts.append("type: procedural")
    parts.append(f"tier: {r.tier}")
    parts.append(f"outcome: {r.outcome}")
    if r.tags:
        parts.append(f"tags: [{', '.join(r.tags)}]")
    parts.append(f"observed: {r.observed_at.strftime('%Y-%m-%dT%H:%M:%SZ')}")
    parts.append("---")
    parts.append("")
    if r.task_context:
        parts.append("## Context")
        parts.append("")
        parts.append(r.task_context)
        parts.append("")
    if r.tool_calls:
        parts.append("## Tool sequence")
        parts.append("")
        parts.append(" -> ".join(r.tool_calls))
    parts.append("")
    return "\n".join(parts)

File: sdks/go/memory/procedural.go
// SPDX-License-Identifier: Apache-2.0

package memory

import (
	"encoding/json"
	"fmt"
	"strings"
	"time"
)

// ProceduralRecord is a structured record of a skill or tool sequence
// invocation, used to build procedural memory over repeated use.
type ProceduralRecord struct {
	Tier        string
	Name        string
	TaskContext string
	Outcome     string
	ObservedAt  time.Time
	ToolCalls   []string
	Tags        []string
}

// DetectProcedurals scans messages for skill invocations (Tier A) and
// agent invocations (Tier B).
func DetectProcedurals(messages []Message) []ProceduralRecord {
	var records []ProceduralRecord
	now := time.Now()

	skills := detectSkillInvocations(messages)
	for _, s := range skills {
		records = append(records, ProceduralRecord{
			Tier:        "skill",
			Name:        s.name,
			TaskContext: s.context,
			Outcome:     s.outcome,
			ObservedAt:  now,
			ToolCalls:   []string{"skill"},
			Tags:        []string{"procedural", "skill", s.name},
		})
	}

	agents := detectAgentInvocations(messages)
	for _, a := range agents {
		records = append(records, ProceduralRecord{
			Tier:        "agent",
			Name:        a.name,
			TaskContext: a.context,
			Outcome:     a.outcome,
			ObservedAt:  now,
			ToolCalls:   []string{"agent"},
			Tags:        []string{"procedural", "agent", a.name},
		})
	}

	return records
}

type invocation struct {
	name    string
	context string
	outcome string
}

func detectSkillInvocations(messages []Message) []invocation {
	var results []invocation

	for i, m := range messages {
		if m.Role != RoleAssistant {
			continue
		}
		for _, tc := range m.ToolCalls {
			if tc.Name != "skill" {
				continue
			}
			var args struct {
				Skill string `json:"skill"`
				Args  string `json:"args"`
			}
			if err := json.Unmarshal(tc.Arguments, &args); err != nil || args.Skill == "" {
				continue
			}

			outcome := inferToolCallOutcome(messages, i, tc.ID)
			ctx := inferProceduralContext(messages, i)

			results = append(results, invocation{
				name:    args.Skill,
				context: ctx,
				outcome: outcome,
			})
		}
	}

	return results
}

func detectAgentInvocations(messages []Message) []invocation {
	var results []invocation

	for i, m := range messages {
		if m.Role != RoleAssistant {
			continue
		}
		for _, tc := range m.ToolCalls {
			if tc.Name != "agent" {
				continue
			}
			var args struct {
				Type   string `json:"type"`
				Prompt string `json:"prompt"`
			}
			if err := json.Unmarshal(tc.Arguments, &args); err != nil || args.Type == "" {
				continue
			}

			outcome := inferToolCallOutcome(messages, i, tc.ID)
			ctx := args.Prompt
			if len(ctx) > 160 {
				ctx = ctx[:160]
			}
			if ctx == "" {
				ctx = inferProceduralContext(messages, i)
			}

			results = append(results, invocation{
				name:    args.Type,
				context: ctx,
				outcome: outcome,
			})
		}
	}

	return results
}

// inferToolCallOutcome looks at the tool result for a given tool call
// to determine outcome.
func inferToolCallOutcome(messages []Message, afterIndex int, toolCallID string) string {
	for _, m := range messages[afterIndex:] {
		if m.Role != RoleTool {
			continue
		}
		if m.ToolCallID == toolCallID || (toolCallID == "" && m.Name == "skill") {
			content := strings.ToLower(m.Content)
			if strings.Contains(content, "error") || strings.Contains(content, "failed") {
				return "error"
			}
			return "ok"
		}
	}
	return "partial"
}

// inferProceduralContext extracts a brief task context from the
// nearest preceding user message.
func inferProceduralContext(messages []Message, beforeIndex int) string {
	for i := beforeIndex - 1; i >= 0; i-- {
		if messages[i].Role == RoleUser {
			ctx := messages[i].Content
			if len(ctx) > 160 {
				ctx = ctx[:160]
			}
			return ctx
		}
	}
	return ""
}

// FormatProceduralRecord formats a procedural record as markdown
// content.
func FormatProceduralRecord(r ProceduralRecord) string {
	var b strings.Builder
	fmt.Fprintf(&b, "---\nname: %s\n", r.Name)
	fmt.Fprintf(&b, "type: procedural\n")
	fmt.Fprintf(&b, "tier: %s\n", r.Tier)
	fmt.Fprintf(&b, "outcome: %s\n", r.Outcome)
	if len(r.Tags) > 0 {
		fmt.Fprintf(&b, "tags: [%s]\n", strings.Join(r.Tags, ", "))
	}
	fmt.Fprintf(&b, "observed: %s\n", r.ObservedAt.Format(time.RFC3339))
	b.WriteString("---\n\n")
	if r.TaskContext != "" {
		fmt.Fprintf(&b, "## Context\n\n%s\n\n", r.TaskContext)
	}
	if len(r.ToolCalls) > 0 {
		fmt.Fprintf(&b, "## Tool sequence\n\n%s\n", strings.Join(r.ToolCalls, " -> "))
	}
	return b.String()
}
